PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp

```diff
@@ -2617,7 +2617,10 @@ void ScintillaEditView::performGlobalStyles()
 	execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_BACK, selectColorBack);
 
 	if (nppParams.isSelectFgColorEnabled())
+	{
 		execute(SCI_SETSELFORE, 1, selectColorFore);
+		execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_TEXT, selectColorFore);

```

Comment:
use the following lines
```
		long newSelectColorFore = selectColorFore;
		newSelectColorFore |= 0xFF000000; // add alpha color to make DirectWrite mode work
		execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_TEXT, newSelectColorFore);
```
instead of 
```
execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_TEXT, selectColorFore);
```
to make both "DirectWrite" and normal mode work

Reply:
*Bravo!* This one-liner also works:

     execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_BACK, (static_cast<long>(selectColorBack) | 0xFF000000));

Reply:
To recap:

* simply passing a 32-bit ARGB color to `SC_ELEMENT_SELECTION_INACTIVE_BACK` resolves the DirectWrite issue (i.e., a solid background overlapping the text), with no regression in GDI mode

#### DirectWrite (with ligatures)

![selectColorFore-D2D](https://user-images.githubusercontent.com/59004801/172090726-e9218cf4-ccda-4665-83d3-fef7a7b0669f.png)

### GDI

![selectColorFore-GDI](https://user-images.githubusercontent.com/59004801/172090731-5440d5cb-e898-41f9-a054-9d7c878b9249.png)

* for the best user experience, we should do the same with `SC_ELEMENT_SELECTION_INACTIVE_TEXT`

The complete revision for approval:

~~~diff
diff --git a/PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp b/PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp
index 59d5eb1e..ea1af2f0 100644
--- a/PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp
+++ b/PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp
@@ -2614,12 +2614,13 @@ void ScintillaEditView::performGlobalStyles()
 		selectColorFore = pStyle->_fgColor;
 	}
 	execute(SCI_SETSELBACK, 1, selectColorBack);
-	execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_BACK, selectColorBack);
+	// pass a 32-bit ARGB color in case DirectWrite is on
+	execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_BACK, (static_cast<long>(selectColorBack) | 0xFF000000));
 
 	if (nppParams.isSelectFgColorEnabled())
 	{
 		execute(SCI_SETSELFORE, 1, selectColorFore);
-		execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_TEXT, selectColorFore);
+		execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_TEXT, (static_cast<long>(selectColorFore) | 0xFF000000));
 	}
 
 	COLORREF caretColor = black;

~~~

Reply:
@Yaron10, @alankilborn ,@donho, just to make sure everyone is on the same page:

1. the one *strictly necessary* change to close the original issue and maintain DirectWrite support is passing a 32-bit ARGB color to `SC_ELEMENT_SELECTION_INACTIVE_TEXT`, like so:

~~~diff
@@ -2588,6 +2588,9 @@ void ScintillaEditView::expand(size_t& line, bool doExpand, bool force, intptr_t
 
 void ScintillaEditView::performGlobalStyles()
 {
+	// used to convert COLORREFs to 32-bit ARGB colors in case DirectWrite is on
+	constexpr unsigned alphaMask = 0xFF000000;
+
 	NppParameters& nppParams = NppParameters::getInstance();
 	const ScintillaViewParams& svp = nppParams.getSVP();
 	
@@ -2618,8 +2621,9 @@ void ScintillaEditView::performGlobalStyles()
 
 	if (nppParams.isSelectFgColorEnabled())
 	{
+		long alphaSelectColorFore = (selectColorFore | alphaMask);
 		execute(SCI_SETSELFORE, 1, selectColorFore);
-		execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_TEXT, selectColorFore);
+		execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_TEXT, alphaSelectColorFore);
 	}
 
 	COLORREF caretColor = black;


~~~

2. However, as [mentioned already](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/11744#issuecomment-1147317777), *every* element that *supports* an alpha channel *should* eventually be set with ARGB colors; not doing so invites undefined behaviour. According to [this table](https://www.scintilla.org/ScintillaDoc.html#SCI_GETELEMENTBASECOLOUR),  __*3*__ of the elements within the scope of `ScintillaEditView::performGlobalStyles()` take ARGB colors:

| SC_ELEMENT_* | Value | Translucent? | 
| :--- | :------ | :----------- | 
|. .  .|. .  .|. .  .|
| SC_ELEMENT_SELECTION_INACTIVE_TEXT| 16 | Translucent | 
| SC_ELEMENT_SELECTION_INACTIVE_BACK| 17 | Translucent | 
|. .  .|. .  .|. .  .|
| SC_ELEMENT_CARET_LINE_BACK | 50 | Translucent | 

So, while *not necessary* to close the original issue, a slightly more complete revision (good enough to satisfy most users until GDI+ is properly implemented) would look something like this:

~~~diff
@@ -2588,6 +2588,9 @@ void ScintillaEditView::expand(size_t& line, bool doExpand, bool force, intptr_t
 
 void ScintillaEditView::performGlobalStyles()
 {
+	// used to convert COLORREFs to 32-bit ARGB colors in case DirectWrite is on
+	constexpr unsigned alphaMask = 0xFF000000;
+
 	NppParameters& nppParams = NppParameters::getInstance();
 	const ScintillaViewParams& svp = nppParams.getSVP();
 	
@@ -2599,7 +2602,8 @@ void ScintillaEditView::performGlobalStyles()
 		pStyle = stylers.findByName(TEXT("Current line background colour"));
 		if (pStyle)
 		{
-			execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_CARET_LINE_BACK, pStyle->_bgColor);
+			long alphaCaretLnBack = (pStyle->_bgColor | alphaMask);
+			execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_CARET_LINE_BACK, alphaCaretLnBack);
 		}
 	}
 
@@ -2613,13 +2617,16 @@ void ScintillaEditView::performGlobalStyles()
 		selectColorBack = pStyle->_bgColor;
 		selectColorFore = pStyle->_fgColor;
 	}
+
+	long alphaSelectColorBack = (selectColorBack | alphaMask);
 	execute(SCI_SETSELBACK, 1, selectColorBack);
-	execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_BACK, selectColorBack);
+	execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_BACK, alphaSelectColorBack);
 
 	if (nppParams.isSelectFgColorEnabled())
 	{
+		long alphaSelectColorFore = (selectColorFore | alphaMask);
 		execute(SCI_SETSELFORE, 1, selectColorFore);
-		execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_TEXT, selectColorFore);
+		execute(SCI_SETELEMENTCOLOUR, SC_ELEMENT_SELECTION_INACTIVE_TEXT, alphaSelectColorFore);
 	}
 
 	COLORREF caretColor = black;

~~~

I have no preference &#x2014; this feature was completely unknown to me before last week.
So: will it be &#x2116; 1, &#x2116; 2, or something else?

Reply:
@rdipardo,

PR closed while I was typing a reply. :)
Thank you for the interesting and enlightening discussion. 👍